File: src/services/emergencyService.ts
// src/services/emergencyService.ts

import api from './api';

// ============================================
// TYPES & INTERFACES
// ============================================

export interface EmergencyServiceType {
  id: string;
  name: string;
  type: 'ambulance' | 'blood' | 'oxygen' | 'doctor' | 'emergency-room' | 'helicopter';
  status: 'available' | 'busy' | 'dispatched' | 'offline' | 'maintenance';
  location: {
    lat: number;
    lng: number;
    address: string;
  };
  eta?: number;
  provider: string;
  phone: string;
  rating?: number;
  availableUnits?: number;
  price?: number;
}

export interface EmergencyRequest {
  id: string;
  requested_by?: string;
  patient_name?: string;
  patient_phone?: string;
  type: 'ambulance' | 'blood' | 'oxygen' | 'doctor';
  priority: 'low' | 'medium' | 'high' | 'critical' | 'urgent' | 'emergency';
  location: string;
  pickup_address?: string;
  status: 'pending' | 'dispatched' | 'picked-up' | 'arrived' | 'completed' | 'cancelled';
  timestamp: string;
  estimatedTime?: number;
  assignedServiceId?: string;
  notes?: string;
  patientId?: string;
  resolvedAt?: string;
  emergency_type?: string;
  needs_oxygen?: boolean;
  hospital_id?: string;
}

export interface CreateEmergencyRequest {
  type: EmergencyRequest['type'];
  priority?: EmergencyRequest['priority'];
  location: string;
  patientId?: string;
  notes?: string;
  contactNumber?: string;
  patient_name?: string;
  patient_phone?: string;
  pickup_address?: string;
  emergency_type?: string;
  needs_oxygen?: boolean;
  hospital_id?: string;
}

export interface BloodDonor {
  id: string;
  name: string;
  bloodGroup: string;
  location: string;
  distance: number;
  phone: string;
  lastDonated?: string;
  available: boolean;
}

export interface OxygenSupplier {
  id: string;
  name: string;
  location: string;
  distance: number;
  price: number;
  stock: string;
  phone: string;
  deliveryTime: string;
  is24Hours: boolean;
}

export interface EmergencyContact {
  id: string;
  name: string;
  relation: string;
  phone: string;
  isPrimary: boolean;
}

export interface EmergencyAlert {
  id: string;
  type: string;
  severity: 'low' | 'medium' | 'high' | 'critical';
  message: string;
  location: string;
  timestamp: string;
  status: 'active' | 'resolved' | 'dispatched';
}

// ============================================
// EMERGENCY SERVICE
// ============================================

export const emergencyService = {
  /**
   * Get all emergency services with optional filters
   */
  getServices: async (type?: string): Promise<EmergencyServiceType[]> => {
    const response = await api.get<any>('/emergency/services', { type });
    if (response.success && response.data) {
      const items = Array.isArray(response.data) ? response.data : response.data?.data || [];
      return items.map((s: any) => ({
        id: String(s.id),
        name: s.name,
        type: s.type,
        status: s.status,
        location: {
          lat: s.latitude || 0,
          lng: s.longitude || 0,
          address: s.address || '',
        },
        eta: s.eta_minutes,
        provider: s.provider || '',
        phone: s.phone || '',
        rating: s.rating,
        availableUnits: s.capacity,
        price: s.price,
      }));
    }
    return [];
  },

  /**
   * Get service by ID
   */
  getServiceById: async (id: string): Promise<EmergencyServiceType> => {
    const response = await api.get<any>('/emergency/services/' + id);
    if (response.success && response.data) {
      const s = response.data;
      return {
        id: String(s.id),
        name: s.name,
        type: s.type,
        status: s.status,
        location: { lat: s.latitude || 0, lng: s.longitude || 0, address: s.address || '' },
        eta: s.eta_minutes,
        provider: s.provider || '',
        phone: s.phone || '',
        rating: s.rating,
        availableUnits: s.capacity,
        price: s.price,
      };
    }
    throw new Error(response.message || 'Emergency service not found');
  },

  /**
   * Get available services by type
   */
  getAvailableServices: async (type?: string): Promise<EmergencyServiceType[]> => {
    const services = await emergencyService.getServices(type);
    return services.filter(s => s.status === 'available');
  },

  /**
   * Create emergency request (ambulance)
   */
  createRequest: async (data: CreateEmergencyRequest): Promise<EmergencyRequest> => {
    const payload: any = {
      patient_name: data.patient_name,
      patient_phone: data.contactNumber || data.patient_phone,
      pickup_address: data.pickup_address || data.location,
      emergency_type: data.emergency_type || data.type,
      needs_oxygen: data.needs_oxygen || false,
      hospital_id: data.hospital_id,
    };

    const response = await api.post<any>('/ambulance/request', payload);
    if (response.success && response.data) {
      const r = response.data;
      return {
        id: r.id,
        requested_by: r.requested_by,
        patient_name: r.patient_name,
        patient_phone: r.patient_phone,
        type: data.type,
        priority: data.priority || 'medium',
        location: r.pickup_address || data.location,
        pickup_address: r.pickup_address,
        status: r.status || 'pending',
        timestamp: r.created_at || new Date().toISOString(),
        estimatedTime: r.estimated_arrival,
        notes: data.notes,
      };
    }
    throw new Error(response.message || 'Failed to create emergency request');
  },

  /**
   * Update emergency request status
   */
  updateRequestStatus: async (id: string, status: EmergencyRequest['status']): Promise<EmergencyRequest> => {
    const response = await api.put<any>(`/ambulance/request/${id}/status`, { status });
    if (response.success && response.data) {
      return response.data;
    }
    throw new Error(response.message || 'Failed to update request status');
  },

  /**
   * Get all emergency requests
   */
  getRequests: async (status?: EmergencyRequest['status']): Promise<EmergencyRequest[]> => {
    const params: any = {};
    if (status) params.status = status;

    const response = await api.get<any>('/ambulance/requests', params);
    if (response.success && response.data) {
      const items = Array.isArray(response.data) ? response.data : response.data?.data || [];
      return items.map((r: any) => ({
        id: r.id,
        requested_by: r.requested_by,
        patient_name: r.patient_name,
        patient_phone: r.patient_phone,
        type: r.emergency_type || 'ambulance',
        priority: r.priority || 'medium',
        location: r.pickup_address || '',
        pickup_address: r.pickup_address,
        status: r.status,
        timestamp: r.created_at,
        estimatedTime: r.estimated_arrival,
        notes: r.notes,
      }));
    }
    return [];
  },

  /**
   * Get emergency request by ID
   */
  getRequestById: async (id: string): Promise<EmergencyRequest> => {
    const response = await api.get<any>('/ambulance/requests/' + id);
    if (response.success && response.data) {
      const r = response.data;
      return {
        id: r.id,
        patient_name: r.patient_name,
        type: r.emergency_type || 'ambulance',
        priority: 'medium',
        location: r.pickup_address || '',
        status: r.status,
        timestamp: r.created_at,
        estimatedTime: r.estimated_arrival,
      };
    }
    throw new Error(response.message || 'Request not found');
  },

  /**
   * Get available ambulances
   */
  getAvailableAmbulances: async (): Promise<EmergencyServiceType[]> => {
    return emergencyService.getAvailableServices('ambulance');
  },

  /**
   * Get all blood donors
   */
  getBloodDonors: async (bloodGroup?: string): Promise<BloodDonor[]> => {
    const params: any = {};
    if (bloodGroup) params.blood_group = bloodGroup;

    const response = await api.get<any>('/blood-donation/donors', params);
    if (response.success && response.data) {
      const items = Array.isArray(response.data) ? response.data : response.data?.data || [];
      return items.map((d: any) => ({
        id: String(d.id),
        name: d.user?.full_name || d.name || 'Donor',
        bloodGroup: d.blood_group,
        location: d.preferred_donation_center || '',
        distance: 0,
        phone: d.user?.phone || '',
        lastDonated: d.last_donation_date,
        available: d.is_available && d.is_eligible,
      }));
    }
    return [];
  },

  /**
   * Search blood donors by location
   */
  searchBloodDonors: async (bloodGroup: string, location: string, maxDistance: number = 5): Promise<BloodDonor[]> => {
    const response = await api.get<any>('/blood-donation/donors/search', {
      blood_group: bloodGroup,
      city: location,
    });
    if (response.success && response.data) {
      const items = Array.isArray(response.data) ? response.data : response.data?.data || [];
      return items.map((d: any) => ({
        id: String(d.id),
        name: d.user?.full_name || d.name || 'Donor',
        bloodGroup: d.blood_group,
        location: d.preferred_donation_center || location,
        distance: 0,
        phone: d.user?.phone || '',
        lastDonated: d.last_donation_date,
        available: d.is_available && d.is_eligible,
      }));
    }
    return [];
  },

  /**
   * Get oxygen suppliers
   */
  getOxygenSuppliers: async (): Promise<OxygenSupplier[]> => {
    const response = await api.get<any>('/oxygen/stocks');
    if (response.success && response.data) {
      const items = Array.isArray(response.data) ? response.data : response.data?.data || [];
      return items.map((s: any) => ({
        id: String(s.id),
        name: s.hospital?.name || s.supplier || 'Oxygen Supplier',
        location: s.hospital?.city || '',
        distance: 0,
        price: 0,
        stock: s.status === 'sufficient' ? 'in-stock' : s.status === 'low' ? 'limited' : 'out-of-stock',
        phone: s.supplier_contact || '',
        deliveryTime: '30 mins',
        is24Hours: !!s.emergency_support,
      }));
    }
    return [];
  },

  /**
   * Get nearby oxygen suppliers
   */
  getNearbyOxygenSuppliers: async (location: string, maxDistance: number = 5): Promise<OxygenSupplier[]> => {
    return emergencyService.getOxygenSuppliers();
  },

  /**
   * Get emergency contacts for current user
   */
  getEmergencyContacts: async (): Promise<EmergencyContact[]> => {
    const response = await api.get<any>('/users/me/emergency-contacts');
    if (response.success && response.data) {
      const items = Array.isArray(response.data) ? response.data : response.data?.data || [];
      if (items.length === 0) {
        // Return default emergency numbers
        return [
          { id: '1', name: 'Emergency Services', relation: 'General', phone: '911', isPrimary: true },
          { id: '2', name: 'Poison Control', relation: 'Medical', phone: '1-800-222-1222', isPrimary: false },
        ];
      }
      return items.map((c: any) => ({
        id: String(c.id),
        name: c.name,
        relation: c.relation,
        phone: c.phone,
        isPrimary: c.is_available,
      }));
    }
    return [
      { id: '1', name: 'Emergency Services', relation: 'General', phone: '911', isPrimary: true },
      { id: '2', name: 'Poison Control', relation: 'Medical', phone: '1-800-222-1222', isPrimary: false },
    ];
  },

  /**
   * Get active emergency announcements
   */
  getActiveEmergencies: async (): Promise<EmergencyAlert[]> => {
    const response = await api.get<any>('/emergency/announcements', { is_active: true });
    if (response.success && response.data) {
      const items = Array.isArray(response.data) ? response.data : response.data?.data || [];
      return items.map((a: any) => ({
        id: String(a.id),
        type: a.type,
        severity: a.priority || 'medium',
        message: a.message,
        location: '',
        timestamp: a.created_at,
        status: 'active',
      }));
    }
    return [];
  },

  /**
   * Cancel emergency request
   */
  cancelRequest: async (id: string, reason?: string): Promise<EmergencyRequest> => {
    return emergencyService.updateRequestStatus(id, 'cancelled');
  },

  /**
   * Get response time statistics
   */
  getResponseStats: async (): Promise<{
    averageResponseTime: number;
    totalRequests: number;
    completedRequests: number;
    criticalResponseTime: number;
  }> => {
    const requests = await emergencyService.getRequests();
    const completed = requests.filter(r => r.status === 'completed');
    const avgTime = completed.reduce((sum, r) => sum + (r.estimatedTime || 0), 0) / (completed.length || 1);

    return {
      averageResponseTime: Math.round(avgTime),
      totalRequests: requests.length,
      completedRequests: completed.length,
      criticalResponseTime: 7,
    };
  },

  /**
   * Subscribe to emergency alerts (placeholder for WebSocket)
   */
  subscribeToAlerts: (callback: (alert: EmergencyAlert) => void): (() => void) => {
    // TODO: Replace with real WebSocket connection when available
    return () => {};
  },
};

// ============================================
// REACT HOOK
// ============================================

export const useEmergencyService = () => {
  return {
    getServices: emergencyService.getServices,
    getServiceById: emergencyService.getServiceById,
    getAvailableServices: emergencyService.getAvailableServices,
    createRequest: emergencyService.createRequest,
    updateRequestStatus: emergencyService.updateRequestStatus,
    getRequests: emergencyService.getRequests,
    getRequestById: emergencyService.getRequestById,
    getAvailableAmbulances: emergencyService.getAvailableAmbulances,
    getBloodDonors: emergencyService.getBloodDonors,
    searchBloodDonors: emergencyService.searchBloodDonors,
    getOxygenSuppliers: emergencyService.getOxygenSuppliers,
    getNearbyOxygenSuppliers: emergencyService.getNearbyOxygenSuppliers,
    getEmergencyContacts: emergencyService.getEmergencyContacts,
    getActiveEmergencies: emergencyService.getActiveEmergencies,
    cancelRequest: emergencyService.cancelRequest,
    getResponseStats: emergencyService.getResponseStats,
    subscribeToAlerts: emergencyService.subscribeToAlerts,
  };
};

export default emergencyService;

File: backend/app/services/emergency_service.py
from sqlalchemy.orm import Session
from sqlalchemy import or_, func
from typing import Optional, List
from datetime import date, datetime

from ..models.emergency import EmergencyService, EmergencyRequest
from ..models.hospital import Hospital
from ..schemas.emergency import (
    EmergencyServiceResponse, EmergencyRequestCreate,
    SOSRequest, EmergencyCancelRequest, EmergencyRequestResponse,
    EmergencyDashboardResponse,
)
from ..core.exceptions import NotFoundException, BadRequestException
from ..utils.helpers import build_pagination_meta, generate_request_number


class EmergencyServiceLayer:
    """Handles all emergency-related business logic."""

    def __init__(self, db: Session):
        self.db = db

    # ============================================
    # CREATE EMERGENCY REQUEST
    # ============================================
    def create_request(self, user_id: str, data: EmergencyRequestCreate) -> EmergencyRequestResponse:
        valid_types = ["ambulance", "blood", "oxygen", "doctor", "emergency-room"]
        if data.type not in valid_types:
            raise BadRequestException(f"Invalid type. Must be one of: {valid_types}")

        valid_priorities = ["low", "medium", "high", "critical"]
        if data.priority not in valid_priorities:
            raise BadRequestException(f"Invalid priority. Must be one of: {valid_priorities}")

        # Try to find nearest available service
        service_id = None
        estimated_time = None
        available_service = (
            self.db.query(EmergencyService)
            .filter(
                EmergencyService.type == data.type,
                EmergencyService.status == "available",
                EmergencyService.is_active == True,
            )
            .first()
        )
        if available_service:
            service_id = available_service.id
            estimated_time = available_service.eta_minutes
            # Mark service as dispatched
            available_service.status = "dispatched"
            available_service.current_load = (available_service.current_load or 0) + 1

        request = EmergencyRequest(
            type=data.type,
            priority=data.priority,
            status="pending",
            patient_id=user_id,
            patient_name=data.patient_name,
            patient_phone=data.patient_phone,
            location_address=data.location_address,
            latitude=data.latitude,
            longitude=data.longitude,
            notes=data.notes,
            service_id=service_id,
            hospital_id=data.hospital_id,
            estimated_time=estimated_time,
        )
        self.db.add(request)

        # If service found, auto-dispatch
        if service_id:
            request.status = "dispatched"

        self.db.commit()
        self.db.refresh(request)
        return EmergencyRequestResponse.model_validate(request)

    # ============================================
    # SOS — IMMEDIATE CRITICAL REQUEST
    # ============================================
    def create_sos(self, user_id: str, data: SOSRequest) -> EmergencyRequestResponse:
        sos_data = EmergencyRequestCreate(
            type=data.type,
            priority="critical",
            patient_name=data.patient_name,
            patient_phone=data.patient_phone,
            location_address=data.location_address,
            latitude=data.latitude,
            longitude=data.longitude,
            notes=data.notes,
        )
        return self.create_request(user_id, sos_data)

    # ============================================
    # GET EMERGENCY STATUS
    # ============================================
    def get_request_status(self, request_id: str) -> EmergencyRequestResponse:
        request = self.db.query(EmergencyRequest).filter(EmergencyRequest.id == request_id).first()
        if not request:
            raise NotFoundException("Emergency request not found")
        return EmergencyRequestResponse.model_validate(request)

    # ============================================
    # CANCEL EMERGENCY REQUEST
    # ============================================
    def cancel_request(self, request_id: str, user_id: str, reason: str) -> EmergencyRequestResponse:
        request = self.db.query(EmergencyRequest).filter(
            EmergencyRequest.id == request_id,
            EmergencyRequest.patient_id == user_id,
        ).first()
        if not request:
            raise NotFoundException("Emergency request not found")

        if request.status in ["completed", "cancelled"]:
            raise BadRequestException(f"Cannot cancel request with status: {request.status}")

        request.status = "cancelled"
        request.cancellation_reason = reason

        # Free up the assigned service
        if request.service_id:
            service = self.db.query(EmergencyService).filter(
                EmergencyService.id == request.service_id
            ).first()
            if service:
                service.status = "available"
                service.current_load = max((service.current_load or 0) - 1, 0)

        self.db.commit()
        self.db.refresh(request)
        return EmergencyRequestResponse.model_validate(request)

    # ============================================
    # GET AVAILABLE AMBULANCES
    # ============================================
    def get_ambulances(self, lat: Optional[float] = None, lng: Optional[float] = None) -> List[EmergencyServiceResponse]:
        query = self.db.query(EmergencyService).filter(
            EmergencyService.type == "ambulance",
            EmergencyService.is_active == True,
        )
        ambulances = query.all()
        return [EmergencyServiceResponse.model_validate(a) for a in ambulances]

    # ============================================
    # GET NEARBY HOSPITALS
    # ============================================
    def get_nearby_hospitals(
        self, lat: Optional[float] = None, lng: Optional[float] = None,
        radius: float = 10, page: int = 1, size: int = 20,
    ) -> dict:
        query = self.db.query(Hospital).filter(
            Hospital.is_active == True,
            Hospital.emergency_service == "active",
        )

        total = query.count()
        hospitals = query.offset((page - 1) * size).limit(size).all()

        items = [
            {
                "id": h.id,
                "name": h.name,
                "type": h.type,
                "city": h.city,
                "state": h.state,
                "phone": h.phone,
                "emergency_phone": h.emergency_phone,
                "latitude": float(h.latitude) if h.latitude else None,
                "longitude": float(h.longitude) if h.longitude else None,
                "available_beds": h.available_beds,
                "icu_available_beds": h.icu_available_beds,
                "emergency_service": h.emergency_service,
                "rating": float(h.rating or 0),
            }
            for h in hospitals
        ]

        return {
            "items": items,
            **build_pagination_meta(total, page, size),
        }

    # ============================================
    # GET EMERGENCY SERVICES
    # ============================================
    def get_services(
        self, type_filter: Optional[str] = None,
        status: Optional[str] = None,
        page: int = 1, size: int = 20,
    ) -> dict:
        query = self.db.query(EmergencyService).filter(EmergencyService.is_active == True)

        if type_filter:
            query = query.filter(EmergencyService.type == type_filter)
        if status:
            query = query.filter(EmergencyService.status == status)

        total = query.count()
        services = query.offset((page - 1) * size).limit(size).all()

        return {
            "items": [EmergencyServiceResponse.model_validate(s) for s in services],
            **build_pagination_meta(total, page, size),
        }

    # ============================================
    # GET USER EMERGENCY REQUESTS
    # ============================================
    def get_user_requests(
        self, user_id: str, status: Optional[str] = None,
        page: int = 1, size: int = 20,
    ) -> dict:
        query = self.db.query(EmergencyRequest).filter(
            EmergencyRequest.patient_id == user_id,
        )

        if status:
            query = query.filter(EmergencyRequest.status == status)

        query = query.order_by(EmergencyRequest.created_at.desc())
        total = query.count()
        requests = query.offset((page - 1) * size).limit(size).all()

        return {
            "items": [EmergencyRequestResponse.model_validate(r) for r in requests],
            **build_pagination_meta(total, page, size),
        }

    # ============================================
    # DASHBOARD
    # ============================================
    def get_dashboard(self) -> EmergencyDashboardResponse:
        active = self.db.query(EmergencyRequest).filter(
            EmergencyRequest.status.in_(["pending", "dispatched", "en-route"]),
        ).count()
        pending = self.db.query(EmergencyRequest).filter(
            EmergencyRequest.status == "pending",
        ).count()
        dispatched = self.db.query(EmergencyRequest).filter(
            EmergencyRequest.status == "dispatched",
        ).count()
        completed_today = self.db.query(EmergencyRequest).filter(
            EmergencyRequest.status == "completed",
        ).count()
        available_ambulances = self.db.query(EmergencyService).filter(
            EmergencyService.type == "ambulance",
            EmergencyService.status == "available",
            EmergencyService.is_active == True,
        ).count()
        nearby_hospitals = self.db.query(Hospital).filter(
            Hospital.is_active == True,
            Hospital.emergency_service == "active",
        ).count()

        return EmergencyDashboardResponse(
            active_requests=active,
            pending_requests=pending,
            dispatched_requests=dispatched,
            completed_today=completed_today,
            available_ambulances=available_ambulances,
            nearby_hospitals=nearby_hospitals,
        )
